### ISSUE TITLE
Can't require node modules in renderer process in v0.20.6 without a refresh

### ISSUE THREAD

**Issue by paulcbetts** [atom/atom-shell#1028]
Our app with v0.20.6 is crashing on first startup, but refreshing the Window fixes it. It appears to be related to module paths:

```
FIRST LOAD:
[  
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/lib/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/node_modules",
  "/private/var/folders/_c/node_modules",
  "/private/var/folders/node_modules",
  "/private/var/node_modules",
  "/private/node_modules",
  "/node_modules",
  "/Users/paul/.node_modules",
  "/Users/paul/.node_libraries",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Frameworks/Slack Helper.app/Contents/lib/node",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/api/lib",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/app",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/common/api/lib"
]

AFTER REFRESH:

[  
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/lib/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/node_modules",
  "/private/var/folders/_c/node_modules",
  "/private/var/folders/node_modules",
  "/private/var/node_modules",
  "/private/node_modules",
  "/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/app/static/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/app/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/node_modules",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/node_modules",
  "/private/var/folders/_c/node_modules",
  "/private/var/folders/node_modules",
  "/private/var/node_modules",
  "/private/node_modules",
  "/node_modules",
  "/Users/paul/.node_modules",
  "/Users/paul/.node_libraries",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Frameworks/Slack Helper.app/Contents/lib/node",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/api/lib",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/app",
  "/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/common/api/lib"
]
```

Note that once we issue a refresh, `$TMPDIR/slack-build/Slack.app/Contents/Resources/app/node_modules` is now in the path list, whereas on first load, it's *not*.

**Comment by paulcbetts** [comment]
Some more info:

Before Refresh:
```
>> require.main

{  
  id:".",
  exports:Object,
  parent:null,
  filename:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h00….app/Contents/Resources/atom/renderer/lib/init.js",
  loaded:true  …
}children:Array[  
  4
]0:Modulechildren:Array[  
  2
]exports:Objectfilename:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/common/lib/init.js"id:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/common/lib/init.js"loaded:trueparent:Modulepaths:Array[  
  14
]__proto__:Module1:Module2:Module3:Modulelength:4__proto__:Array[  
  0
]exports:Objectfilename:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/lib/init.js"id:"."loaded:trueparent:nullpaths:Array[  
  14
]0:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/lib/node_modules"1:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/renderer/node_modules"2:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/atom/node_modules"3:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/node_modules"4:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/node_modules"5:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/node_modules"6:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/node_modules"7:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/node_modules"8:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/node_modules"9:"/private/var/folders/_c/node_modules"10:"/private/var/folders/node_modules"11:"/private/var/node_modules"12:"/private/node_modules"13:"/node_modules"length:14__proto__:Array[  
  0
]__proto__:Module
}
```

After Refresh:

```
>>> require.main

{  
  id:".",
  exports:Object,
  parent:null,
  filename:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h00…lack.app/Contents/Resources/app/static/index.html",
  loaded:true  …
}children:Array[  
  9
]exports:Objectfilename:"/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Resources/app/static/index.html"id:"."loaded:trueparent:nullpaths:Array[  
  27
]__proto__:Module
}
```

**Comment by zcbenz** [comment]
@paulcbetts Can you also print the following variables at the beginning of app:

* `window.location.protocol`
* `process.argv`

**Comment by paulcbetts** [comment]
Before:

```
>> window.location.protocol
file:

>> ["/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Frameworks/Slack Helper.app/Contents/MacOS/Slack Helper", "--type=renderer", "--js-flags=--harmony", "--no-sandbox", "--enable-deferred-image-decoding", "--lang=English", "--node-integration=true", "--subpixel-font-scaling=false", "--enable-delegated-renderer", "--enable-impl-side-painting", "--num-raster-threads=1", "--disable-accelerated-video-decode", "--channel=6140.1.277471217"]
```

After:

```
>> window.location.protocol
file:

>> process.argv
["/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Frameworks/Slack Helper.app/Contents/MacOS/Slack Helper", "--type=renderer", "--js-flags=--harmony", "--no-sandbox", "--enable-deferred-image-decoding", "--lang=English", "--node-integration=true", "--subpixel-font-scaling=false", "--enable-delegated-renderer", "--enable-impl-side-painting", "--num-raster-threads=1", "--disable-accelerated-video-decode", "--channel=6140.5.40585886"]
```

**Comment by zcbenz** [comment]
I can not find out why this happened, it seems that the [code to inject module search paths](https://github.com/atom/atom-shell/blob/master/atom/renderer/lib/init.coffee#L59-L72) didn't execute, but according to your output this should not happen. Can you debug locally whether the code in the link has executed?

**Comment by paulcbetts** [comment]
@zcbenz Is there a trick to break in super early and trace through that stuff? Does the node-debugger stuff work still?

**Comment by zcbenz** [comment]
Using [node-inspector to debug](https://github.com/atom/atom-shell/blob/master/docs/tutorial/debugging-browser-process.md) used to work, but with recent updates of V8 it is not working now. I'm afraid you have to use `console.log` to trace now.

**Comment by paulcbetts** [comment]
@zcbenz It looks like at the time that code is executing, `window.location` points to "about:blank"

**Comment by paulcbetts** [comment]
It looks like this bug is a race condition in the renderer between Chromium and `NodeBindings::CreateEnvironment`, where we're trying to access `window.location` too early, before we've actually started to load the page.

**Comment by zcbenz** [comment]
The node environment is created in `DidCreateScriptContext` in atom-shell, it seems that the page is not guaranteed to load at this point, I'll move the initialization code to `DidCreateDocumentElement` and see if it works.

**Comment by zcbenz** [comment]
I found that this happens when the user sends a message to web contents immediately after loading url, like this:

```javascript
myWindow.loadUrl('http://github.com');
myWindow.webContents.send('message');
```

It will also influence `<webview>` under some conditions.

**Comment by zcbenz** [comment]
Fixed by #1044.
